### PR TITLE
Separate out constant node index information from ExecutionFrame

### PR DESCRIPTION
**Pull request by skottmckay**
- Move constant node index information into SessionState, and create once after any transforms are done.
- Avoid calls to copy to/from devices in ExecuteQuery if there is only the CPU provider 
  - simplistic initial optimization for CPU only environments
  - these checks for whether a copy is needed start to show up in profiling with the changes to ExecutionFrame, even though no copy is required
  - the copy to/from device logic was recently added to subgraph execution to handle when the subgraph runs on GPU
    - there are further optimizations that can be done there as well - first version was to always copy to CPU when running the control flow node. we need to add checks to avoid that copy where possible.
  - can also be smarter if there are multiple execution providers by tracking if any device copies were needed for the input and outputs
    - will make those changes as a separate PR

### CODE REVIEW

File: include/onnxruntime/core/graph/graph.h [hunk 1]
```diff
@@ -283,8 +283,12 @@ class Node {
   void ToProto(ONNX_NAMESPACE::NodeProto& proto) const;
 
   /** Call the provided function for all explicit inputs, implicit inputs, and outputs of this Node.
-      If the NodeArg is an explicit or implicit input, is_input will be true when func is called. */
-  void ForEachDef(std::function<void(const onnxruntime::NodeArg&, bool is_input)> func) const;
+      If the NodeArg is an explicit or implicit input, is_input will be true when func is called. 
+	  @param include_missing_optional_defs Include NodeArgs that are optional and were not provided
```
Comment (KeDengMS) [inline include/onnxruntime/core/graph/graph.h]: >	 [](start = 0, length = 1)

nit: tabs. Please use .clangformat

File: include/onnxruntime/core/graph/graph.h [hunk 2]
```diff
@@ -283,8 +283,12 @@ class Node {
...
 
   /** Call the provided function for all explicit inputs, implicit inputs, and outputs of this Node.
-      If the NodeArg is an explicit or implicit input, is_input will be true when func is called. */
-  void ForEachDef(std::function<void(const onnxruntime::NodeArg&, bool is_input)> func) const;
+      If the NodeArg is an explicit or implicit input, is_input will be true when func is called. 
+	  @param include_missing_optional_defs Include NodeArgs that are optional and were not provided 
+	                                       i.e. NodeArg::Exists() == false.
+	  */
+  void ForEachDef(std::function<void(const onnxruntime::NodeArg&, bool is_input)> func,
+                  bool include_missing_optional_defs = false) const;
```
Comment (KeDengMS) [inline include/onnxruntime/core/graph/graph.h]: >include_missing_optional_defs [](start = 23, length = 29)

How about OnlyIterateExistingDefs with default of True?

File: onnxruntime/core/framework/node_index_info.cc [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        int index;
+        Status status = mlvalue_idx_map.GetIdx(name, index);
+        ORT_ENFORCE(status.IsOK(), status.ErrorMessage());
+        node_values_[cur_idx] = index;
+      }
+      // else it's a missing optional input or output so leave the -1
+
+      ++cur_idx;
+    },
+                    include_missing_optional_defs);
```
Comment (KeDengMS) [inline onnxruntime/core/framework/node_index_info.cc]: >                     [](start = 0, length = 20)

indent here seems confusing. Maybe try .clangformat?

File: onnxruntime/core/framework/execution_providers.h [hunk 4]
```diff
@@ -76,6 +76,8 @@ class ExecutionProviders {
 
   bool Empty() const { return exec_providers_.empty(); }
 
+  size_t NumProviders() const { return exec_providers_.size(); }
```
Comment (KeDengMS) [inline onnxruntime/core/framework/execution_providers.h]: >NumProviders [](start = 9, length = 12)

maybe size() or count() matches better with begin()/end() below.

File: onnxruntime/core/framework/node_index_info.cc [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  for (const auto& node : graph_viewer.Nodes()) {
+    node.ForEachDef([&](const onnxruntime::NodeArg& /*arg*/, bool /*is_input*/) {
+      ++total_def_count;
+    },
+                    include_missing_optional_defs);
+  }
+
+  // init all to -1
+  node_offsets_.resize(graph_viewer.MaxNodeIndex(), -1);
```
Comment (KeDengMS) [inline onnxruntime/core/framework/node_index_info.cc]: >-1 [](start = 52, length = 2)

Could be more explicit with a enum of InvalidNodeIndex?

File: include/onnxruntime/core/graph/graph.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/graph/graph.h]: Weird. I do use clangformat on all saves so not sure where the tabs came from.

---
In reply to: [252390544](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252390544) [](ancestors = 252390544)

File: include/onnxruntime/core/graph/graph.h (hunk 2 above)
Comment (skottmckay) [reply include/onnxruntime/core/graph/graph.h]: exclude_missing_optional_defs?

---
In reply to: [252392502](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252392502) [](ancestors = 252392502)

File: onnxruntime/core/framework/node_index_info.cc (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/framework/node_index_info.cc]: That is clangformat. The indent is aligning with the opening '(' of the ForEachDef call. The lambda which is the first argument has different indent rules which is why it looks weird. 

---
In reply to: [252393442](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252393442) [](ancestors = 252393442)

File: onnxruntime/core/framework/execution_providers.h (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/framework/execution_providers.h]: The reason for 'begin' and 'end' is so you can use this class in a range based loop. Those functions don't match our naming convention though, but it's what C++ requires. Adding more things that don't match the naming convention starts to make it more confusing as to what the convention is, so I deliberately didn't use 'size' or 'count'. 

---
In reply to: [252394402](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252394402) [](ancestors = 252394402)

File: include/onnxruntime/core/graph/graph.h (hunk 2 above)
Comment (KeDengMS) [reply include/onnxruntime/core/graph/graph.h]: Was stumbled with the concept of implicit/explicit/optional input at the header, but after checking graph.cc I think the name is fine.

---
In reply to: [252437302](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252437302) [](ancestors = 252437302,252392502)

File: onnxruntime/core/framework/node_index_info.cc (hunk 3 above)
Comment (KeDengMS) [reply onnxruntime/core/framework/node_index_info.cc]: Maybe you can move the lambda in ForEachDef in a new line to make it align better for reading.

---
In reply to: [252437805](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252437805) [](ancestors = 252437805,252393442)

File: onnxruntime/core/graph/graph.cc [hunk 6]
```diff
@@ -550,19 +550,20 @@ const Graph* Node::GetGraphAttribute(const std::string& attr_name) const {
   return const_cast<Node*>(this)->GetMutableGraphAttribute(attr_name);
 }
 
-void Node::ForEachDef(std::function<void(const onnxruntime::NodeArg&, bool is_input)> func) const {
+void Node::ForEachDef(std::function<void(const onnxruntime::NodeArg&, bool is_input)> func,
+                      bool include_missing_optional_defs) const {
   for (const auto* arg : InputDefs()) {
-    if (arg->Exists())
+    if (include_missing_optional_defs || arg->Exists())
```
Comment (pranavsharma) [inline onnxruntime/core/graph/graph.cc]: Not sure how this change relates to the execution frame optimization? for my education - why is this needed?

File: onnxruntime/core/framework/execution_frame.h [hunk 7]
```diff
@@ -94,11 +93,8 @@ class ExecutionFrame {
     return all_values_[mlvalue_index];
   }
 
-  // Index to the first argument of the given node.
-  int GetFirstArgIndex(onnxruntime::NodeIndex index) const {
-    ORT_ENFORCE(index < node_offsets_.size());
-    return node_offsets_[index];
-  }
+  // Get the index for the first entry of the given node.
+  int GetNodeOffset(onnxruntime::NodeIndex index) const;
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.h]: ah, good. thanks for making this name obvious :-)

File: onnxruntime/core/framework/execution_frame.h [hunk 8]
```diff
@@ -128,9 +124,6 @@ class ExecutionFrame {
  private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(ExecutionFrame);
 
-  // This method is not thread safe!
-  void Release(int offset);
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.h]: is this not called anywhere?

File: onnxruntime/core/framework/session_state.cc [hunk 9]
```diff
@@ -206,4 +207,21 @@ const SessionState* SessionState::GetSubgraphSessionState(onnxruntime::NodeIndex
   return session_state;
 }
 
+void SessionState::CalculateNodeIndexInfo() {
+  ORT_ENFORCE(graph_viewer_);
+  node_index_info_ = std::make_unique<NodeIndexInfo>(*graph_viewer_, mlvalue_name_idx_map_);
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/framework/session_state.cc]: Creation is delayed until the Graph is fully transformed and resolved as nodes may be removed. 

---
In reply to: [252499727](https://github.com/Microsoft/onnxruntime/pull/410#discussion_r252499727) [](ancestors = 252499727)

File: onnxruntime/core/framework/utils.cc [hunk 10]
```diff
@@ -218,7 +218,6 @@ common::Status MatchOutputsWithProviders(const SessionState& session_state,
 
     for (auto* arg : node.OutputDefs()) {
       if (!arg->Exists() ||
-          arg->Name().empty() ||
           !(found = Contains(output_names, arg->Name())).first) {
```
Comment (skottmckay) [inline onnxruntime/core/framework/utils.cc]: Same value as Exists so no need to call

File: onnxruntime/core/framework/utils.cc [hunk 11]
```diff
@@ -258,44 +257,6 @@ common::Status MatchOutputsWithProviders(const SessionState& session_state,
     }
   }
 
-  // If we've already seen all the outputs requested just return.
-  if (seen_outputs.size() == output_names.size()) {
-    return Status::OK();
-  }
-
-  // Handle the case when a constant is an output but has been folded into a weight
-  // and hence it doesn't show up in any of the OutputDefs before.
```
Comment (skottmckay) [inline onnxruntime/core/framework/utils.cc]: Handling by changing the order in ExecutionFrame::Init instead.

File: onnxruntime/core/framework/execution_frame.h [hunk 12]
```diff
@@ -60,30 +62,6 @@ class ExecutionFrame {
...
-  // This tensor does not own the buffer.
-  // The executor / planner need to be careful about the
-  // lifetime of the buffer. Tensor itself won't manage it.
-  // This method is not thread safe!
-  Status AllocateTensorWithPreAllocateBuffer(int offset,
-                                             void* pBuffer,
-                                             MLDataType element_type,
-                                             const OrtAllocatorInfo& location,
-                                             const TensorShape& shape);
-
```
Comment (skottmckay) [inline onnxruntime/core/framework/execution_frame.h]: These were only called from unit tests. Update unit tests (short term) to use AllocateMLValue... versions
